// html-validate/dist/cjs/meta-helper.js
'use strict';

var utils_naturalJoin = require('./utils/natural-join.js');

/**
 * Helper function to assist IDE with completion and type-checking.
 *
 * @public
 */
function defineMetadata(metatable) {
    return metatable;
}

/**
 * @internal
 */
function allowedIfAttributeIsPresent(...attr) {
    return (node) => {
        if (attr.some((it) => node.hasAttribute(it))) {
            return null;
        }
        const expected = utils_naturalJoin.naturalJoin(attr.map((it) => `"${it}"`));
        return `requires ${expected} attribute to be present`;
    };
}
/**
 * @internal
 */
function allowedIfAttributeIsAbsent(...attr) {
    return (node) => {
        const present = attr.filter((it) => node.hasAttribute(it));
        if (present.length === 0) {
            return null;
        }
        const expected = utils_naturalJoin.naturalJoin(present.map((it) => `"${it}"`));
        return `cannot be used at the same time as ${expected}`;
    };
}
/**
 * @internal
 */
function allowedIfAttributeHasValue(key, expectedValue, { defaultValue } = {}) {
    return (node) => {
        const attr = node.getAttribute(key);
        if (attr && typeof attr !== "string") {
            return null;
        }
        const actualValue = attr ? attr : defaultValue;
        if (actualValue && expectedValue.includes(actualValue.toLocaleLowerCase())) {
            return null;
        }
        const expected = utils_naturalJoin.naturalJoin(expectedValue.map((it) => `"${it}"`));
        return `"${key}" attribute must be ${expected}`;
    };
}
/**
 * @internal
 */
function allowedIfParentIsPresent(...tags) {
    return (node) => {
        const match = tags.some((it) => node.closest(it));
        if (match) {
            return null;
        }
        const expected = utils_naturalJoin.naturalJoin(tags.map((it) => `<${it}>`));
        return `requires ${expected} as parent`;
    };
}
/**
 * @public
 */
const metadataHelper = {
    allowedIfAttributeIsPresent,
    allowedIfAttributeIsAbsent,
    allowedIfAttributeHasValue,
    allowedIfParentIsPresent,
};

exports.defineMetadata = defineMetadata;
exports.metadataHelper = metadataHelper;
